casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala

```diff
@@ -119,38 +126,35 @@ class RhoSpec(
       } yield runtime
 
       runtimeResource.use { runtime =>
-        for {
-          _    <- logger.info("Starting tests from " + testObject.path)
-          _    <- setupRuntime(runtime)
-          rand = Blake2b512Random(128)
-          _ <- TestUtil
-                .eval(testObject.code, runtime, testObject.normalizerEnv)(
-                  implicitly,
-                  rand.splitShort(1)
-                )
-                .timeout(executionTimeout)
-
-          result <- testResultCollector.getResult
-        } yield result
-      }
+        evaluateDeploy(rhoSpecDeploy, runtime) >> tests.toList
+          .traverse_ {
+            case (source, userSk) =>
+              ConstructDeploy.sourceDeployNowF[Task](
+                source,
+                sec = userSk
+              ) >>= (evaluateDeploy(_, runtime))
+          }
+          .timeout(executionTimeout)
+      } >> testResultCollector.getResult
     }
 
-  private def setupRuntime(runtime: Runtime[Task]): Task[Runtime[Task]] =
-    for {
-      _ <- evalDeploy(rhoSpecDeploy, runtime)
-      // reset the deployParams.userId before executing the test
-      // otherwise it'd execute as the deployer of last deployed contract
-      _ <- runtime.deployParametersRef.update(_.copy(userId = Par()))
-    } yield runtime
-
-  private def evalDeploy(
-      deploy: DeployData,
-      runtime: Runtime[Task]
-  ): Task[Unit] = {
+  private def evaluateDeploy(deploy: DeployData, runtime: Runtime[Task]): Task[Unit] = {
     implicit val rand: Blake2b512Random = Blake2b512Random(
       ProtoUtil.stripDeployData(deploy).toProto.toByteArray
     )
-    eval(deploy.term, runtime, NormalizerEnv(deploy.toProto))
+    implicit val c: _cost[Task] = runtime.cost
+    runtime.deployParametersRef.set(ProtoUtil.getRholangDeployParams(deploy)) >> Interpreter[Task]
+      .injAttempt(
+        runtime.reducer,
+        runtime.errorLog,
+        deploy.term,
+        Cost.UNSAFE_MAX,
+        NormalizerEnv(deploy.toProto)
+      ) >>= {
+      case EvaluateResult(_, errors) =>
+        if (errors.nonEmpty) errors.head.raiseError[Task, Unit]
```

Comment:
only the first error is raised and the others are hidden. You concatenate the error messages and raise an error containing all the failure information

Reply:
My thinking is, if we're running a test that consists of multiple programs, the test should terminate the first time an error is thrown. If a program fails, the programs after it are likely to fail as well.

Reply:
I agree with your reasoning but I'm thinking that the errors field is a List for a reason. 

Or do you think it should be changed (in a future PR)?